cmd/process-agent/collector.go

```diff
@@ -92,24 +97,83 @@ func (l *Collector) runCheck(c checks.Check, payloads chan checkPayload) {
 	messages, err := c.Run(l.cfg, atomic.AddInt32(&l.groupID, 1))
 	if err != nil {
 		log.Errorf("Unable to run check '%s': %s", c.Name(), err)
-	} else {
-		payloads <- checkPayload{messages, c.Name()}
-		// update proc and container count for info
-		updateProcContainerCount(messages)
-		if !c.RealTime() {
-			d := time.Since(s)
-			switch {
-			case runCounter < 5:
-				log.Infof("Finished %s check #%d in %s", c.Name(), runCounter, d)
-			case runCounter == 5:
-				log.Infof("Finished %s check #%d in %s. First 5 check runs finished, next runs will be logged every 20 runs.", c.Name(), runCounter, d)
-			case runCounter%20 == 0:
-				log.Infof("Finish %s check #%d in %s", c.Name(), runCounter, d)
-			}
+		return
+	}
+
+	for _, m := range messages {
+		body, err := encodePayload(m)
+		if err != nil {
+			log.Errorf("Unable to encode message: %s", err)
+			continue
+		}
+
+		extraHeaders := make(http.Header)
+		extraHeaders.Set(api.HostHeader, l.cfg.HostName)
+		extraHeaders.Set(api.ProcessVersionHeader, Version)
+		extraHeaders.Set(api.ContainerCountHeader, strconv.Itoa(getContainerCount(m)))
+
+		if cid, err := clustername.GetClusterID(); err == nil && cid != "" {
+			extraHeaders.Set(api.ClusterIDHeader, cid)
+		}
+
+		if err := l.acquireQueueBytes(len(body)); err != nil {
+			log.Errorf("Unable to enqueue message: %s", err)
+			continue
+		}
+
+		payload := checkPayload{
+			name:    c.Name(),
+			body:    body,
+			headers: extraHeaders,
+		}
+		select {
+		case payloads <- payload:
+		default:
+			bytesEnqueued := atomic.LoadInt32(&l.bytesEnqueued)
+			log.Warnf("Delivery queue is full.  Size=%d, Bytes Enqueued=%d.  Waiting for space...", len(payloads), bytesEnqueued)
+
+			payloads <- payload
+		}
+	}
+
+	// update proc and container count for info
+	updateProcContainerCount(messages)
+	if !c.RealTime() {
+		d := time.Since(s)
+		switch {
+		case runCounter < 5:
+			log.Infof("Finished %s check #%d in %s", c.Name(), runCounter, d)
+		case runCounter == 5:
+			log.Infof("Finished %s check #%d in %s. First 5 check runs finished, next runs will be logged every 20 runs.", c.Name(), runCounter, d)
+		case runCounter%20 == 0:
+			log.Infof("Finish %s check #%d in %s", c.Name(), runCounter, d)
 		}
 	}
 }
 
+func (l *Collector) acquireQueueBytes(bytesRequested int) error {
+	max := int32(l.cfg.QueueBytes)
+
+	for {
+		currentSize := atomic.LoadInt32(&l.bytesEnqueued)
+		newSize := int32(bytesRequested) + currentSize
+		if newSize > max {
+			return fmt.Errorf("cannot acquire %d bytes of queue space, queue size %d would exceed capacity %d", bytesRequested, newSize, max)
+		}
+		if atomic.CompareAndSwapInt32(&l.bytesEnqueued, currentSize, newSize) {
+			break
+		}
+	}
+
+	return nil
+}
+
+func (l *Collector) releaseQueueBytes(bytesReleased int) {
+	if v := atomic.AddInt32(&l.bytesEnqueued, -1*int32(bytesReleased)); v < 0 {
+		log.Errorf("Releasing %d bytes from queue caused value to go negative: %d.", bytesReleased, v)
```

Comment:
Do we understand in what circumstances can this happen? If this happens, should we force a recalculation of `bytesEnqueued`?

Reply:
It should not happen at all, I added this for defensive debugging purposes after the fact in case a bug did occur (or perhaps was introduced at a later date)

I thought about resetting it but given that I'm unaware of any current circumstances where this could occur, I could not think of a good way to remediate the value so I left it alone.